proto-grpc.md

```diff
@@ -0,0 +1,254 @@
+# Protobuf and gRPC with Bazel
+
+The use of [Protobuf](https://developers.google.com/protocol-buffers) and
+[gRPC](https://grpc.io) with [Bazel](https://bazel.build) is a particularly common scenario, since
+these three technologies provide a complimentary base of build, serialisation and remote-call upon
+which to build complex systems and products. This document summarises the current state of the
+Protobuf/gRPC Bazel ecosystem and gives advice on getting started when you are looking to integrate
+one of these tools into your project.
+
+Given that all three of Bazel, gRPC and Protobuf are Google-derived projects, you may expect that
+there is a simple integration between them. However, the existence of this document should indicate
+that there is not a single simple answer to integrating these tools. Originally, Protobuf and gRPC
+rules were baked into Bazel for a few key languages, but this has understandably not continued to
+scale with the many new langauges that are now supported in external Bazel rulesets. As a result,
+independent rulesets have been released by the community that provide the integrations necessary to
+convert `.proto` files into a wide variety of languages and libraries.
+
+This page is expected to continue as a 'live' document; as the ecosystem evolves and new rulesets
+are released, this document can be edited at
+[SIG-rules-authors](https://github.com/bazel-contrib/SIG-rules-authors/tree/gh-pages) to continue to
+reflect today's state.
+
+
+## The Basics
+
+At its most basic level, the task of working with Protobuf and gRPC `.proto` files is much the same
+as elsewhere: we execute the Protobuf Compiler `protoc` with the necessary arguments, environment
+and plugins to get it to produce a desired set of files or libraries for any specific language. For
+those unfamiliar with protoc, it is a small compiler executable that parses sets of `.proto`
+formatted files and executes plugins to write a set of output source code files. These files
+implement the interfaces for working with the messages and services in the language supported by the
+plugin.
+
+However, given the prevalence of Protobuf and gRPC use, users are not expected to reinvent the rules
+needed for calling protoc themselves. Instead, the common base dependency ruleset
```

Comment:
I think I'd cut this section at "there's a base rule called `proto_library` and then language-specific rules that depend on it"

Then have a dedicated section in the doc for `proto_library` and rules_proto and the bazel-gazelle plugin for generating those.

Maybe even saying "Step 1 in any Bazel project with proto and grpc is to accurately generate the `proto_library` rules, here's how to do that and verify you're in a good state before moving on"

Reply:
We can't rely on having Gazelle for auto-generation of proto_library targets in this intro, as most of the rulesets listed later in the doc do not provide this tooling. As far as I'm aware, only stackb/rules_proto and rules_go have this ability. All the other rulesets rely on users manually writing these targets.

I have restructured this section a bit to better integrate the bazel query example and point out that some rulesets do make this easier.

Reply:
wait, can't everyone generate the `proto_library` targets? Agree they'll have to manually write `mylang_proto_library` targets next to them, but in many cases they only need those for a subset of the `proto_library` targets, and at least the former gives you the right dep graph for the proto descriptors.

Reply:
In principle, yes I believe you could run Gazelle with the 'file' only mode and it should output `proto_library` targets. However, I'm not sure you can isolate it from making go related changes when doing so? This is also quite a big step for a getting started guide, since we now have to introduce the concept of Gazelle and it is perhaps more useful to start from first principles of "you just need a `proto_library`".

Perhaps a section on 'Working with Larger Codebases' or 'Automating Rules' where we introduce this as a useful advancement on the manual way in the getting started (+xref)? Since most small to moderate codebases aren't really going to need to introduce Gazelle for the few proto files they have (if they aren't using Go that is)

Reply:
gazelle definitely does do proto without any Go. I think if gazelle were easy to use, we'd tell most users to use it, even in a small/medium codebase you don't want to gain enough bazel expertise to know how to write all the BUILD files by hand I think. But we can have it in another section and/or add it later yes.